envoy/stream_info/stream_info.h

```diff
@@ -375,20 +375,80 @@ class DownstreamTiming {
   void addWireBytesReceived(uint64_t added_bytes) { wire_bytes_received_ += added_bytes; }
 
+  struct BytesSnapshot {
+    SystemTime snapshot_time;
+    uint64_t header_bytes_sent{};
+    uint64_t header_bytes_received{};
+    uint64_t wire_bytes_sent{};
+    uint64_t wire_bytes_received{};
+  };
+  void takeDownstreamPeriodicLoggingSnapshot(const SystemTime& snapshot_time) {
+    downstream_periodic_logging_bytes_snapshot_ = std::make_unique<BytesSnapshot>();
+
+    downstream_periodic_logging_bytes_snapshot_->snapshot_time = snapshot_time;
+    downstream_periodic_logging_bytes_snapshot_->header_bytes_sent = header_bytes_sent_;
+    downstream_periodic_logging_bytes_snapshot_->header_bytes_received = header_bytes_received_;
+    downstream_periodic_logging_bytes_snapshot_->wire_bytes_sent = wire_bytes_sent_;
+    downstream_periodic_logging_bytes_snapshot_->wire_bytes_received = wire_bytes_received_;
+  }
+  void takeUpstreamPeriodicLoggingSnapshot(const SystemTime& snapshot_time) {
+    upstream_periodic_logging_bytes_snapshot_ = std::make_unique<BytesSnapshot>();
+
+    upstream_periodic_logging_bytes_snapshot_->snapshot_time = snapshot_time;
+    upstream_periodic_logging_bytes_snapshot_->header_bytes_sent = header_bytes_sent_;
+    upstream_periodic_logging_bytes_snapshot_->header_bytes_received = header_bytes_received_;
+    upstream_periodic_logging_bytes_snapshot_->wire_bytes_sent = wire_bytes_sent_;
+    upstream_periodic_logging_bytes_snapshot_->wire_bytes_received = wire_bytes_received_;
+  }
+  const BytesSnapshot* bytesAtLastDownstreamPeriodicLog() const {
+    return downstream_periodic_logging_bytes_snapshot_.get();
+  }
+  const BytesSnapshot* bytesAtLastUpstreamPeriodicLog() const {
+    return upstream_periodic_logging_bytes_snapshot_.get();
+  }
+  // Adds the bytes from `existing` to `this`.
+  // Additionally, captures periodic bytes meters from `existing`, adds them to `this`, and merges
+  // bytes from `this` into the periodic bytes meters. This is because the existing periodic bytes
+  // meter (especially the downstream periodic meter) may already have bytes accumulated and a
+  // create_time_ set on it.
+  void captureExistingBytesMeter(BytesMeter& existing) {
+    // Add bytes accumulated on `this` to the pre-existing periodic bytes collectors.
+    if (existing.downstream_periodic_logging_bytes_snapshot_) {
+      downstream_periodic_logging_bytes_snapshot_ =
+          std::move(existing.downstream_periodic_logging_bytes_snapshot_);
+      existing.downstream_periodic_logging_bytes_snapshot_ = nullptr;
+    }
+    if (existing.upstream_periodic_logging_bytes_snapshot_) {
+      upstream_periodic_logging_bytes_snapshot_ =
+          std::move(existing.upstream_periodic_logging_bytes_snapshot_);
+      existing.upstream_periodic_logging_bytes_snapshot_ = nullptr;
+    }
+
+    // Accumulate existing bytes.
+    header_bytes_sent_ += existing.header_bytes_sent_;
```

Comment:
I may be missing something here but isn't there a chance that ``header_bytes_sent_`` will be accumulated twice since ``addHeaderBytesSent`` also increases its count?

Reply:
There is no more call to addHeaderBytesSent in setUpstreamBytesMeter -- I moved all that incrementing to inside this function instead.

Reply:
Yeah but there are other locations in the filters site that would call those functions, for example:
https://github.com/envoyproxy/envoy/blob/26797819a34cc4461d8cf2c0019463ab4d04509f/source/common/tcp_proxy/tcp_proxy.cc#L670

Isn't the accumulation will occur twice in such case? (where there's also a periodic log)

Reply:
the usecase for this function is "capture `existing` and then replace `existing` with `this`" functionality.

i.e. `existing` has seen multiple calls to `addWireBytesReceived`, and now existing is going to be replaced by `this`, which needs to know about all the previous bytes received. 

Example timeline with random numbers:
**T0**: downstream periodic log occurs for `existing` bytes meter
- existing->wire_bytes_received_ = 10
- existing->downstream_logging_snapshot_wire_bytes_ = 10

**T1**: 10 additional wire bytes are received:
- existing->wire_bytes_received_ = 20
- existing->downstream_logging_snapshot_wire_bytes_ = 10

**T2**: bytes meters get swapped `existing` for `new`. New already has 5 bytes received on it. 
- new->wire_bytes_received_ = 25
- new->downstream_logging_snapshot_wire_bytes_ = 10
- existing->wire_bytes_received_ = 20
- existing->downstream_logging_snapshot = nullptr.
existing is now dropped, we won't increment it anymore.

**T3**: 6 more wire bytes arrive. They increment the `new` bytes meter.
- new->wire_bytes_received_ = 31
- new->downstream_logging_snapshot_wire_bytes_ = 10

**T4**: a new downstream periodic log occurs. Since the last log, there have been 21 bytes. Nothing has been double counted.